Add shift(): wrapper for front and popFront

wilzbach opened the pull request:
Hey all, 

I am quite new to D and I was wondering why there is no convenience wrapper which combines `front` and `popFront`. I know there is `takeOne` (which doesn't modify the source) and `dropOne` (which will return the resulting array after dropping one element).

I think the best example of such a method in other languages is the `next` method in [Python](https://docs.python.org/3/library/functions.html#next), thus the naming. However feel free to suggest other names. Other ideas would be `shift`, `takeFirst`, `removeFirst`, `shiftFirst`.
If it makes sense to you I am happy to add the opposite method (`takeLast`,`removeLast`,`shiftLast`).

Pros:
- syntactic sugar (see below for an example)
- supported natively in [most](http://rosettacode.org/wiki/Queue/Usage) languages
- I learned stack syntax with return at university (it's even on [wikipedia](https://en.wikipedia.org/wiki/Stack_(abstract_data_type))

```d
auto myFile = ["<crazy header>","1","2"];
// cut off the header
myFile.next.writeln;
// do sth with the body
myFile.map!(to!int).map!"a + 1".map!(to!string).joiner(",").writeln;
```

Cons:
- yet another method
- possible side effects?

As mentioned above I am just starting with D, so let me know if there is a better way to write this PR or you would add more checks (afaik `front` and `popFront` will throw an Assertion Error if the range is empty, so this behavior should be induced).

I initially asked this question at the DLang forum, which lead me to submit this PR.
https://forum.dlang.org/post/orcjkabkcffueduqkyml@forum.dlang.org

greenify commented:
Update: there's another argument to add this method - there seems to be justified confusion that `moveFront` is the convenience wrapper ([example 1](https://dlang.org/library/std/range/primitives/move_front.html), [example 2](https://forum.dlang.org/post/na0jbj$2bji$1@digitalmars.com))

JakobOvrum commented:
This comes up from time to time, but the main issue is that `popFront` can often invalidate the return value of `front`, that is; the value of `front` is often only valid until the next call to `popFront`. We've sometimes called these ranges "transient ranges". A prominent example is File.byLine.

I think it's also a small issue in practice, as higher level range algorithms can take care of calling the range primitives correctly.

e10s commented:
FYI: https://issues.dlang.org/show_bug.cgi?id=14598

quickfur commented:
I think I prefer the names `frontPop` from issue 14598. `next` sounds a bit too generic for my liking, and seems too prone to name-coliisions with local variables and aggregate members.

greenify commented:
> that is; the value of front is often only valid until the next call to popFront.

@JakobOvrum: Can't we then duplicate the front value as this method is really about convenience. 

> I think I prefer the names `frontPop`

Wouldn't it be a bit confusing to have `frontPop` and `popFront`? Do you prefer it over `shiftFirst` or `takeFirst`? 

Any opinion from @MartinNowak?

quickfur commented:
@greenify Of course we can duplicate `.front`, _if_ we knew it was a transient range... the problem is, how to know this in generic code? In general, you can't tell.

Having said that, though, transient ranges aren't very common, so most of the time this _should_ work. And it *does* indeed simplify certain kinds of code, e.g., recursive descent parsers, where the pattern `if (r.front.match(...)) { r.popFront(); doSomething; }` occurs frequently.

jmdavis commented:
The reality of the matter is that ranges with a transient front tend to break if you do much more than use `foreach` on them. They don't even work with `std.array.array`. So, I would not consider the fact that this won't work with transient front to be an issue at all, particularly since it's not like you can really fix that. This fundamentally doesn't work with transient front, just like plenty of other range-based stuff.

As for the name, I think that `frontPop` and `backPop` are very descriptive of what they're doing and definitely better than `next`, which _will_ conflict with plenty of code. `take` has a very different meaning in std.range, so it would not make any sense to use something with that in its name, and `shift` isn't a concept that we have with ranges, so I expect that plenty of folks would be confused by it. So, while there might be some risk of confusion between `frontPop` and `popFront`, it seems like the best choice out of the currently suggested options at least.

greenify commented:
> As for the name, I think that frontPop and backPop are very descriptive of what they're doing and definitely better than next

I renamed `next` to `frontPop` and added `backPop` too - any opinions on the code or tests?

JackStouffer commented on std/range/primitives.d:
```diff
@@ -1987,6 +1987,81 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+    return e;
+}
+
+@safe unittest
```
Examples should come from ddoc-ed unit tests, not in documentation. All you have to do is add `///` to the previous line

JackStouffer commented on std/range/primitives.d:
```diff
@@ -1987,6 +1987,81 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+assert(r == [2,3])
+---
+ */
+auto frontPop(Range)(ref Range r)
```
Why `ref`?

quickfur replied:
`ref` is needed so that by-value ranges will work as expected. Otherwise odd things may happen.

JackStouffer commented on std/range/primitives.d:
```diff
@@ -1987,6 +1987,81 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
 /**
+Pops the first element from $(D r) and returns it.
+It combines $(D front) and $(D popFront) in one method.
+
```
`Params` and `Returns` sections needed here. See other functions in this file for examples

JackStouffer commented on std/range/primitives.d:
```diff
@@ -1987,6 +1987,81 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+/**
+Pops the last element from $(D r) and returns it.
+It combines $(D back) and $(D popBack) in one method.
+
```
Same comments about unittest examples and documentation sections

greenify commented:
thanks a lot @JackStouffer :)

> Examples should come from ddoc-ed unit tests, not in documentation. All you have to do is add `///` to the previous line

Thanks! I didn't know that - I added the `///` magic to all four test or should it be only be added to first test of a function (I found both cases in `primitives.d`)?

> Why ref?

Because otherwise (as far as I understood it) the pointers to the range will be copied and popping will only modify the copy - see also e.g. `popFrontN`

> Params and Returns sections needed here. See other functions in this file for examples

I could only find one place in `primitives.d` (the module documentation) that defines `Params` - is this a new convention?

-> Added & squashed.

JackStouffer commented:
> Thanks! I didn't know that - I added the /// magic to all four test or should it be only be added to first test of a function (I found both cases in primitives.d)?

The secondary unit tests don't add any extra value, so I don't think that they should be ddoc-ed.

>I could only find one place in primitives.d (the module documentation) that defines Params - is this a new convention?

They were added a while ago. All of the functions in Phobos should have them, but a lot haven't been updated yet.

These functions could also use the following contract

```
auto frontPop(Range)(ref Range r)
if (isInputRange!Range)
in
{
    assert(!r.empty, "r cannot be empty");
}
body
{
    auto e = r.front;
    r.popFront();
    return e;
}
```

greenify commented:
> The secondary unit tests don't add any extra value, so I don't think that they should be ddoc-ed.

Okay.

> These functions could also use the following contract

I thought that `popFront`/`popBack` has this contract too and will throw it up throw the stack. However this makes it more obvious to the user. Thanks again :)

JackStouffer commented:
LGTM

quickfur commented:
Please add links to `frontPop` and `backPop` in the comment headers in `std/range/primitives.d` as well as `std/range/package.d`.

greenify commented:
@quickfur added to header in `std/range/primitives.d`, but I couldn't find a place in `std/range/package.d`. Could you give a more explicit point there? Thanks!

quickfur commented:
Hmm, I guess `std/range/package.d` doesn't have that table anymore. I guess that means you don't have to worry about it. :-)

quickfur commented:
LGTM

BBasile commented on std/range/primitives.d:
```diff
@@ -1987,6 +1993,93 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+if (isInputRange!Range)
+in
+{
+    assert(!r.empty, "r cannot be empty");
```
The assertion message is not useful (the __r__). Either it must be removed or changed for more value added.
I suggest `assert(!r.empty, "empty range passed in " ~ __PRETTY_FUNCTION__); `. With pretty function, the type of the argument is also written.

same remark for `backPop` input contract.

greenify commented:
@BBasile wow thanks - I didn't know that special keyword existed!

quickfur commented on std/range/primitives.d:
```diff
@@ -1987,6 +1993,93 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+if (isBidirectionalRange!Range)
+in
+{
+    assert(!r.empty, "empty range passed in " ~ __PRETTY_FUNCTION__);
```
Shouldn't the message be "... passed *to*" rather than "passed in"?

quickfur commented on std/range/primitives.d:
```diff
@@ -1987,6 +1993,93 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+if (isInputRange!Range)
+in
+{
+    assert(!r.empty, "empty range passed in " ~ __PRETTY_FUNCTION__);
```
ditto

greenify commented:
@quickfur - you are right: `to` is grammatically correct. Updated ;-)

kyllingstad commented:
One thing which has not been mentioned yet is the fact that these functions are not exception safe, and they never can be. That is, if `popFront()` throws, then `front` is forever lost. (Maybe the `return` statement can throw too, if the compiler for some reason decides it cannot apply RVO.)

Therefore, I vote against putting these functions in Phobos. They are trivial enough that anyone can easily add them to their own code in a matter of seconds anyway if they accept this drawback.

Another option is to mark them as `nothrow`, so they can only be used if `front`'s postblit and `popFront()` are also `nothrow`.

quickfur commented:
ping @andralex 
Need a decision here.

schveiguy commented:
I would vote no on this enhancement. `popFront` may invalidate whatever `front` returned, and a range doesn't have to guarantee anything in terms of the lifetime of what `front` returns beyond the next call to `popFront`. For instance `File.byLine`.

I see others have brought this up. This would cause no end of problems IMO.

If you ordered them `popFront`, then `front` (with a check for `empty` throwing?), it could work.

quickfur commented:
Unfortunately `isTransient` can only be implemented if the range implementor defines its value. There is no way to know whether a range is transient or not just by looking at its type and the signatures of `front`, `popFront`, etc..  Having said that, though, since transient ranges are generally the exception rather than the norm, we *could* adopt a convention that transient ranges declare themselves thus by defining a `.transient` enum member or something like that, and have `isTransient` return `false` if a range doesn't have such a member.

On a related note, there are a good number of Phobos algorithms that *do not work* on transient ranges at all, because internally they (blindly) assume that assigning the value of `.front` to a local variable will retain its original value after `popFront` is called. I've fixed a couple of places where this happens, but I'm almost certain many other places haven't been fixed yet, and in some cases, *cannot* be fixed without restricting the algorithms to forward ranges only rather than input ranges. Some things simply cannot be implemented without copying `.front` somehow, and the only way to do this reliably with the current API is to require forward ranges and use `save` to keep track of the previous position of the range.

And just a sidenote, calling `.dup` is not a generic solution, because transient ranges may return non-array types that point to a common buffer space that gets reused by `popFront`. You can only use `.dup` in user code where you know the concrete type of the element; in Phobos generic algorithms we can't know this. And besides, even `.dup` won't help you if `.front` is an array of references to a common buffer. You can `.dup` it all you want and it still won't prevent invalidation when `popFront` is called.

Anyway, this whole issue with transient ranges is something that was overlooked in the initial design of ranges, because originally, ranges were conceived to be abstractions of iterators, or "views" into some underlying data. At the time, data generators that reuse state with no actual underlying data weren't really on the radar. So nobody even thought of the idea of `popFront` invalidating previous values of `front`. Most of the ranges at the time weren't transient, so this issue wasn't even known until things like `File.byLine` came along, and, AFAICT, originally started as a non-transient range that `.dup`'d each line, but later on due to poor performance was changed to reuse the line buffer, thus becoming transient.  By then, a good number of Phobos algorithms had been written that assumed that it was OK to copy `.front` to a local variable and call `popFront`. I'm pretty sure many of those algorithms still exist today, and still exhibit breakage when used with transient ranges.  I've been meaning to get around to fixing them, but just haven't been able to keep up, what with all the other things going on with D.

quickfur commented:
P.S. A further comment about `.dup`: even if there are no nested references, calling `.dup` isn't always the right thing to do. For example, what if the whole point of the range is to return an array of references to something? In that case it would be wrong to copy the referents just because we're trying to be transient-safe.  Basically, you cannot solve this in a generic way without the author of the range offering additional information on how the range was meant to be used.

And even exporting a `.transient` member isn't the final solution to this issue either -- consider what you should do if, say, you have a chain of ranges `r.filter!somePred`. The resulting range is a `Filter` wrapping the original range `r`. But if `r` is transient, then the `Filter` wrapper needs to export this fact as well, otherwise, any downstream range algorithm that needs to know if something is transient will fail to detect this fact, and malfunction. But that means that *every* range algorithm must become aware of `.transient`, even if the algorithm itself doesn't care whether something is transient or not. IOW, this solution is too invasive and requires too much boilerplate in every single range algorithm, and all this just so that we can work with transient ranges, which are, generally speaking, pretty rare.

So this issue is very complicated albeit rare, and it's questionable where we draw the line of proposed solutions costing more than the ROI.

schveiguy commented:
> there are a good number of Phobos algorithms that do not work on transient ranges at all

Right, but this isn't an algorithm, it's a primitive, or a wrapper of primitives. It's dead simple to avoid (just call `popFront` after you are done with `front`).

> something that was overlooked in the initial design of ranges

I disagree. The fact that forward ranges exist proves they were not overlooked. If you have some notion of a forward range, then you have some notion of a non-forward range, which means the data is gone once it's not in `front` (you can't go back).

I think what happened is that algorithms that work fine on some ranges don't work fine on others. And this is just a fact of life.

One thing that *could* work:

```D
auto frontPop(Range)(ref Range r) nothrow
if (isForwardRange!Range)
in
{
    assert(!r.empty, "empty range passed to " ~ __PRETTY_FUNCTION__);
}
body
{
    auto e = x.save.front;
    r.popFront();
    return e;
}
```

I do not know of any existing expectation that when a range goes out of scope, it would destroy any values you retrieved via front, but this may be the only viable option for something like this. For most cases, the call to `save` will be a noop.

greenify commented:
@schveiguy 
> One thing that could work:

yes this seems to work nicely - I added a test with a simple transient range that overwrites is array element :)
If I understood it correctly then `nothrow` is no longer required as `save` will create a copy, so `front` won't be gone even if `popFront` throws an error.

> For most cases, the call to save will be a noop.

I like it -  nice solution!

jmdavis commented on std/range/primitives.d:
```diff
@@ -1987,6 +1993,177 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+if (isForwardRange!Range)
+in
+{
+    assert(!r.empty, "empty range passed to " ~ __PRETTY_FUNCTION__);
```
An assertion without a message should be plenty. It's downright silly to have to put __PRETTY_FUNCTION__ in an assertion message. The `AssertError` already has that information. If the problem is that stacktraces don't show the function names, then that is what needs to be fixed rather than trying to work around it all over the place.

jmdavis commented on std/range/primitives.d:
```diff
@@ -1987,6 +1993,177 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+}
+body
+{
+    auto e = r.save.front;
```
This is just downright weird and merits a comment. Also, if the issue is ranges with a transient `front`, as I recall, the conclusion of previous discussions on the topic was that Phobos isn't going to support them for anything except simple input ranges (and even that was iffy). If they happen to work, great, but too many algorithms simply won't work with them, and having to call `save` just because you want to call `front` is horrible. And the rest of Phobos definitely doesn't do that.

So, if we're doing it, then there should be a comment explaining it, but I'd strongly vote against calling `save`. It makes it so that this function doesn't work with input ranges just because of concern over a particular range type that already causes havoc with most algorithms anyway.

schveiguy replied:
What we would need is a version that handles forward ranges that return reference-containing types, and one that handles input ranges that return value types.

jmdavis commented:
Personally, I don't think that this function is worth adding, much is might occasionally be useful - particularly when you take the issues that have already been discussed into account (like exception safety). It just doesn't do enough, and there are too many caveats with using it. But I guess that that's up to @andralex, since we obviously can't agree on it.

schveiguy commented:
@jmdavis I agree, I think the value this function adds is quite small.

greenify commented:
Wow this was a very active discussion and I learned a lot about D :)

>Personally, I don't think that this function is worth adding, much is might occasionally be useful - particularly when you take the issues that have already been discussed into account (like exception safety). It just doesn't do enough, and there are too many caveats with using it.

I do understand this too now - I will close it and let's move on. Thanks again for the discussion and your arguments/concerns.

greenify commented:
FYI: I found `backPop` in Phobos!

https://github.com/D-Programming-Language/phobos/blob/master/std/net/isemail.d#L1958

It's defined as `private`, so one can't access it - but given this discussion it's funny it that was already part of Phobos :D

CyberShadow commented:
> Personally, I don't think that this function is worth adding

I disagree.

Other languages that have this idiom (remove and return first element from collection):

- JavaScript has `.shift()`
- Python has `.pop(0)`
- Perl has `shift`
- Ruby has `.shift`
- Rust has `.pop`
- Lua has `table.remove(t,1)`

Most of the above are just for arrays (or equivalent). The implementation here works with both arrays and ranges.

I have found use for such a function many times, generally for arrays.

@wilzbach I suggest:

- Rename the function to `shift`, as it is what it is called most often in other languages.
- Make it explicit that it works with arrays as well as ranges.
- Document the caveat regarding exceptions.

ntrel commented:
Maybe call this `consumeFront`. There's a discussion (from 2012) about whether combining front, popFront can be more efficient:
http://forum.dlang.org/post/mailman.92.1341428120.31962.digitalmars-d@puremagic.com

Even if the unicode decoding doesn't benefit much (I only skimmed the thread), it seems in theory other ranges might do.

I don't see that restricting this to forward ranges is worthwhile - you prevent transient ranges but you also add invisible inefficiency. It's overkill to save a whole range when you may only need the current element. With a name like `consumeFront` you could argue it's obvious transient ranges shouldn't be used.

@kyllingstad: 
> these functions are not exception safe, and they never can be. That is, if popFront() throws, then front is forever lost. (Maybe the return statement can throw too, if the compiler for some reason decides it cannot apply RVO.)

Good point, I think `r.popFront` should have to be `nothrow` but I think it doesn't matter if `front` throws.

wilzbach commented on std/range/primitives.d:
```diff
@@ -1987,6 +1993,136 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+ */
+auto shift(Range)(ref Range r)
+if (isInputRange!Range
+    && functionAttributes!(popFrontN!Range) & FunctionAttribute.nothrow_)
```
Not a typo - `popFrontN` does the attribute inference for us - just `popFront` wouldn't work due to the difference between attribute and function.

wilzbach replied:
> We could use a private alias if that's too cryptic.

Updated to use one - it's more readable for the user this way ;-)

wilzbach commented:
> I have found use for such a function many times, generally for arrays.

Wohoo - apparently I am not the only one missing this function.

> @wilzbach I suggest:

done.

> Good point, I think r.popFront should have to be nothrow but I think it doesn't matter if front throws.

done.

> don't see that restricting this to forward ranges is worthwhile 

Removed and added a warning about transient ranges.

> Maybe call this consumeFront.

Good idea, what do I other people think?
(I personally think that we should try to have a similar naming to other languages).

CyberShadow commented:
> (I personally think that we should try to have a similar naming to other languages).

Considering that it is very unlikely that D would be the first language where a user encounters such a function, I think it would be more useful to use the names existing in other languages, barring any strong reasons such as conflicts (which I don't see).

ZombineDev commented on std/range/primitives.d:
```diff
@@ -1986,6 +1992,140 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+}
+
+///
+@safe pure unittest
```
Please add `@nogc` and `nothrow`. For `@nogc` you need to make the following change:
```
int[3] arr = [1, 2, 3];
int r[] = arr[];
```

wilzbach replied:
Thanks & added.
Even though I think this makes the first test for the user a bit uglier.

ZombineDev replied:
Thanks.
Yes, but it shows an interesting (for newbies) effect: only the slice gets modified and not the original array, which IIRC is not the case in other languages like JS.
Maybe it's worth adding `assert (arr.equals(4.iota))` at the end.

wilzbach replied:
> Maybe it's worth adding assert (arr.equals(4.iota)) at the end.

done ;-)

ZombineDev commented on std/range/primitives.d:
```diff
@@ -1986,6 +1992,140 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+    assert(r.empty);
+}
+
+@safe pure unittest
```
Ditto here and for the other exampmes.

ZombineDev commented on std/range/primitives.d:
```diff
@@ -1986,6 +1992,146 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
+    Next element of the input range
+ */
+auto shift(Range)(ref Range r)
+if (isInputRange!Range && hasNothrowPopFront!Range)
```
Nitpick: please indent the `if` with 4 spaces. Ditto for `shiftBack`.

wilzbach replied:
hmm after having a short look at the other template constraints in `primitives` , they also don't seem to use indention

ZombineDev replied:
I'm not sure where you looked. The only exception is `putChar`:
```
$ pcregrep -MHn "\)\nif" range/primitives.d
range/primitives.d:252:correct primitive: $(D r.put(e)) if $(D R) defines $(D put), $(D r.front = e)
if $(D r) is an input range (followed by $(D r.popFront())), or $(D r(e))
range/primitives.d:348:private void putChar(R, E)(ref R r, E e)
if (isSomeChar!E)
```

```
$ pcregrep -MHn "\)\n    if" range/primitives.d
range/primitives.d:1596:auto walkLength(Range)(Range range)
    if (isInputRange!Range && !isInfinite!Range)
range/primitives.d:1610:auto walkLength(Range)(Range range, const size_t upTo)
    if (isInputRange!Range)
range/primitives.d:1667:size_t popFrontN(Range)(ref Range r, size_t n)
    if (isInputRange!Range)
range/primitives.d:1703:size_t popBackN(Range)(ref Range r, size_t n)
    if (isBidirectionalRange!Range)
range/primitives.d:1798:void popFrontExactly(Range)(ref Range r, size_t n)
    if (isInputRange!Range)
range/primitives.d:1814:void popBackExactly(Range)(ref Range r, size_t n)
    if (isBidirectionalRange!Range)
```

```
$ pcregrep -MHn "\)\n    if" range/package.d | wc -l
86
$ pcregrep -MHn "\)\n    if" algorithm/* | wc -l
170
```

CyberShadow replied:
Andrei seems to prefer the current style.

https://github.com/dlang/phobos/pull/2011#discussion_r57523495

ZombineDev replied:
You mean Andrei prefers the **less** common style. OK, can't argue with that.

wilzbach replied:
>  The only exception is putChar

```
> grep "^if (" primitives.d
if (r.empty) {}   // can test for empty
if (isSomeChar!E)
if (isInputRange!Range && hasNothrowPopFront!Range)
if (isBidirectionalRange!Range && hasNothrowPopBack!Range)
if (!isNarrowString!(T[]) && !is(T[] == void[]))
if (isNarrowString!(C[]))
if (!isNarrowString!(T[]) && !is(T[] == void[]))
if (isNarrowString!(T[]))
if (!isNarrowString!(T[]) && !is(T[] == void[]))
if (!isNarrowString!(T[]))
```

wilzbach replied:
> You mean Andrei prefers the less common style. OK, can't argue with that.

Guys, I don't really care about this - the main point here is to get `shift` into Phobos :)
Btw we should just enable some code style tool like [DScanner](https://github.com/Hackerpilot/Dscanner) for Phobos and agree once on one of both ;-)

ZombineDev replied:
OK, I agree.

ZombineDev commented on std/range/primitives.d:
```diff
@@ -1988,6 +1994,156 @@ ElementType!R moveAt(R)(R r, size_t i)
+    import std.algorithm.comparison: equal;
+    import std.range: iota;
+    // original array stays unmodified
+    assert((cast(int[]) arr).equal(iota(1, 4)));
```
How about: `assert(arr[].equal(iota(1, 4)));`

wilzbach replied:
thanks - that's a lot nicer :)

ZombineDev commented:
Overall LGTM, but I think we have to do something about ranges with non-copyable elelements to improve the API symmetry:

| function  | action                  | wrapper around                                    |
|-----------:|----------------------:|---------------------------------------------------:|
| `shift`   | copy front + advance | `auto e = r.front(); r.popFront(); return e;`     |
| `?`       | move front + advance | `auto e = r.moveFront(); r.popFront(); return e;` |
| `dropOne` |       [void] + advance | `r.popFront(); return r;`                         |

wilzbach commented:
> Overall LGTM, but I think we have to do something about ranges with non-copyable elelements to improve the API symmetry:

Couldn't we call `moveFront` in shift if it's not copyable?

ZombineDev commented:
This won't work with elements can be both moved and copied - e.g. copying a `RefCounted` element will increase it's ref count, moving it will not. Users should be able to use the semantics the fit their use case.

ZombineDev commented:
If we don't come up with a good name, we can always add a template parameter:
```
auto shift(bool moveInsteadOfCopying = false, Range)(ref Range r)
{
    static if (!isCopyable!(ElementType!Range) && !moveInsteadOfCopying)
        static assert (0, "Can't copy elements with disabled postblit! Use `shift!true` instead!");
    // ...
}
```

ntrel commented:
> move front + advance

I think this is quite a bit less common so it's less clear a wrapper is needed for that.

ZombineDev commented:
Yes, maybe for scripts you are right. However when your application is `main() @nogc`, Phobos' lacking support for move-only types (like smart pointers) becomes a huge problem. You start hitting problems every step of the way and the only solution is to reimplement everything yourself, because even simple things, such as `OutputRange`s, don't work.

D is a great systems programming language,  but unfortunately Phobos was not written and is not actively developed with `@nogc` in mind. (By `@nogc` I mean use in constraint environments in general, `version (FreeStanding)`, etc., not just avoiding the GC). Phobos shortcommings look like D shortcomings to people new to D developing for constrained environments, which is bad and untrue. 

Ranges are great in helping you avoid allocations and unnecessary work, but that works until your application grows large enough that resource management becomes unavoidable. Then suddenly the lacking move support becomes a deal breaker.

We can't leave things as they are. We need to start somewhere.

andralex commented:
I'm 75/25 against this.

+ Other languages have it, some with the same name
+ Helps eliminate possible confusion over `moveFront`
- Enshrines wrong semantics for transient input ranges. That alone carries a lot of weight.
- Forces return by value even when front yields a reference.
- It can be done in two lines of code, and does not save the user from learning the range API.
- Loss of information: if popFront throws, the popped value is lost so there's no shift for those range. Another view on this matter is the abstraction is just not that good.
- Although other languages have it, their range abstraction is considerably less flexible so they can afford a more rigid semantics.

The way I look at it there's just too much cons to it. There is real value to the function, but it's weighed against powerful issues.

I do see a few positives coming out of this:

* Reopened the discussion of transient ranges. 
* Good learning experience for the submitter - this is one pass through the whole gamut of writing/documenting/unittesting/community discussion.

I'll close this, but don't take it as standoffish. If anyone finds a pro argument that they think is killer, please reopen.

nordlow commented:
I would like to see a new trait `isTransientRange` to safely enable this for the ranges that have non-destructive `popFront`. Could this be expressed somehow using a new range primitive?

andralex commented:
@nordlow thanks for your input. I'd say by the point the explanation is 5x the size of the two-liner, it's better to do without. Never mind `{steal,take,pull}{Front,Back}` etc. Go and do great things.
